common/transport/amqp/src/amqp.ts

```diff
@@ -145,44 +235,190 @@ export class Amqp {
+            debug('In the disconnect handler for the connecting session state with error of: ' + this._getErrorName(context.connection.error));
+            let callback = this._sessionCallback;
+            this._sessionCallback = undefined;
+            manageConnectionHandlers('removeListener');
+            this._fsm.transition('disconnected', callback, new errors.NotConnectedError('rhea: connection disconnected'));
+          },
+          amqpError: (err) => {
+            this._fsm.transition('disconnecting', null, err);
           },
-          amqpError: (err) => this._fsm.transition('disconnecting', null, err),
-          amqpDisconnected: () => debug('ignoring disconnected event while connecting'),
           '*': () => this._fsm.deferUntilTransition()
         },
         connected: {
           _onEnter: (connectCallback, result) => {
             /*Codes_SRS_NODE_COMMON_AMQP_16_002: [The `connect` method shall establish a connection with the IoT hub instance and if given as argument call the `done` callback with a null error object in the case of success and a `results.Connected` object.]*/
             this._safeCallback(connectCallback, null, new results.Connected(result));
           },
-          amqpError: (err) => this._fsm.transition('disconnecting', null, err),
-          amqpDisconnected: () => this._fsm.transition('disconnected', undefined, new errors.NotConnectedError('amqp10: connection closed')),
+          'session_error': (context: EventContext) => {
+            debug('In the session_error handler for the connected state with error of: ' + this._getErrorName(context.session.error));
+            debug('Deferring transition until the session_close event occurs.');
+            this._indicatedSessionError = context.session.error;
+          },
+          'session_close': (context: EventContext) => {
+            debug('In the session_close handler for the connected state with indicated error of: ' + this._getErrorName(this._indicatedSessionError));
+            debug('Transition to disconnecting');
+            let err = this._indicatedSessionError;
+            this._indicatedSessionError = undefined;
+            this._sessionCloseOccurred = true;
+            this._fsm.transition('disconnecting', null, err);
+          },
+          'connection_error': (context: EventContext) => {
+            debug('In the connection_error handler for the connected state with error of: ' + this._getErrorName(context.connection.error));
+            debug('Deferring transition until the connection_close occurs.');
+            this._indicatedConnectionError = context.connection.error;
+          },
+          'connection_close': (context: EventContext) => {
+            debug('In the connection_close handler for the connected state with indicated error of: ' + this._getErrorName(this._indicatedConnectionError));
+            debug('Transition to disconnecting');
+            let err = this._indicatedConnectionError;
```

Comment:
>et err = this._indicatedConnectionError; [](start = 13, length = 40)

Similar vague uncomfortable feeling.  Does this assume that connection_error always happens before connection_close if you're in this state?

Reply:
No.  See comment above.

---
In reply to: [209389010](https://github.com/Azure/azure-iot-sdk-node/pull/344#discussion_r209389010) [](ancestors = 209389010)